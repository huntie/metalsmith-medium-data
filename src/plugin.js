import { take } from 'lodash';
import fetch from 'node-fetch';
import { getPosts } from './selectors';

export const DEFAULT_OPTIONS = {
  key: 'medium',
  limit: 10
};

export default function plugin(userOptions) {
  const options = { ...DEFAULT_OPTIONS, ...userOptions };

  return async (files, metalsmith, done) => {
    if (!options.username) {
      done(new Error('A Medium username must be provided.'));

      return;
    }

    try {
      const response = await fetch(`https://medium.com/@${options.username}/latest`, {
        headers: { Accept: 'application/json' }
      });
      const { payload } = JSON.parse((await response.text()).replace(/^[^{]*/, ''));

      Object.assign(metalsmith.metadata(), {
        [options.key]: {
          user: payload.user,
          posts: take(getPosts(payload), options.limit)
        }
      });
    } catch (e) {
      done(new Error('Failed to fetch data from the Medium API.'));

      return;
    }

    done();
  };
}
