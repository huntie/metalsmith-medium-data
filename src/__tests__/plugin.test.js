import fetch from 'node-fetch';
import plugin, { DEFAULT_OPTIONS } from '../plugin';

jest.mock('node-fetch');

const testOptions = { username: 'huntie' };

describe('plugin', () => {
  let metadata;
  let metalsmith;

  beforeEach(() => {
    fetch.mockImplementation(() =>
      Promise.resolve({
        text: jest.fn(() =>
          Promise.resolve(
            // Reconstruct API response body received including unwanted prefix
            '])}while(1);</x>' +
              JSON.stringify(require('./sample-response.json'))
          )
        )
      })
    );

    metadata = {};
    metalsmith = {
      metadata: jest.fn(() => metadata)
    };
  });

  test('should call done() when run', async () => {
    const done = jest.fn();

    await plugin(testOptions)({}, metalsmith, done);

    expect(done).toHaveBeenCalledTimes(1);
  });

  test('should call done() with error if username option not set', async () => {
    const done = jest.fn();

    await plugin({})({}, metalsmith, done);

    expect(done.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          [Error: A Medium username must be provided.],
        ],
      ]
    `);
  });

  test('should call done() with error on network error', async () => {
    const done = jest.fn();
    fetch.mockImplementation(() => Promise.reject());

    await plugin(testOptions)({}, metalsmith, done);

    expect(done.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          [Error: Failed to fetch data from the Medium API.],
        ],
      ]
    `);
  });

  test('should assign mapped response object to key on metadata', async () => {
    await plugin(testOptions)({}, metalsmith, () => {});

    expect(metadata).toMatchSnapshot();
  });

  test('should assign to specified key on metadata if option set', async () => {
    await plugin({ ...testOptions, key: 'cms' })({}, metalsmith, () => {});

    expect(Object.keys(metadata)).toMatchInlineSnapshot(`
      Array [
        "cms",
      ]
    `);
  });

  test('should limit post entries to number set by limit option', async () => {
    await plugin({ ...testOptions, limit: 2 })({}, metalsmith, () => {});

    expect(metadata[DEFAULT_OPTIONS.key].posts).toHaveLength(2);
  });
});
