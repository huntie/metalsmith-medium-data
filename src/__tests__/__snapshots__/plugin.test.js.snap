// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin should assign mapped response object to key on metadata 1`] = `
Object {
  "medium": Object {
    "posts": Array [
      Object {
        "canonicalUrl": "",
        "coverless": true,
        "createdAt": 1551977936172,
        "creatorId": "e54e8fe3cbbe",
        "detectedLanguage": "en",
        "displayAuthor": "",
        "firstPublishedAt": 1551980533590,
        "inResponseToPostId": "",
        "inResponseToRemovedAt": 0,
        "isNsfw": false,
        "isProxyPost": false,
        "isSeries": false,
        "latestPublishedAt": 1552597402774,
        "latestPublishedVersion": "a1d43b7635ab",
        "license": 0,
        "previewImage": Object {
          "backgroundSize": "",
          "filter": "",
          "height": 0,
          "imageId": "1*r6kRETnqirT230S4nVwJRQ.png",
          "originalHeight": 1317,
          "originalWidth": 2044,
          "strategy": "resample",
          "width": 0,
        },
        "readingTime": 1.260377358490566,
        "recommends": 2,
        "responsesCreatedCount": 0,
        "slug": "why-im-all-in-with-citymapper-pass",
        "subtitle": "A short one, but I felt like sharing some thoughts about the recently announced Citymapper Pass in London, and why I’ve picked one up…",
        "tags": Array [
          Object {
            "name": "Travel",
            "slug": "travel",
          },
          Object {
            "name": "London Underground",
            "slug": "london-underground",
          },
          Object {
            "name": "Tfl",
            "slug": "tfl",
          },
          Object {
            "name": "Lifehacks",
            "slug": "lifehacks",
          },
        ],
        "title": "Why I’m all in with Citymapper Pass",
        "totalClapCount": 2,
        "type": "Post",
        "url": "https://medium.com/@huntie/why-im-all-in-with-citymapper-pass-832f905ce92a",
        "versionId": "a1d43b7635ab",
      },
      Object {
        "canonicalUrl": "",
        "coverless": true,
        "createdAt": 1546888584419,
        "creatorId": "e54e8fe3cbbe",
        "detectedLanguage": "en",
        "displayAuthor": "",
        "firstPublishedAt": 1547149706637,
        "inResponseToPostId": "",
        "inResponseToRemovedAt": 0,
        "isNsfw": false,
        "isProxyPost": false,
        "isSeries": false,
        "latestPublishedAt": 1547662856305,
        "latestPublishedVersion": "e59120a0fc7",
        "license": 0,
        "previewImage": Object {
          "backgroundSize": "",
          "filter": "",
          "height": 0,
          "imageId": "1*1l7iPfcwUt4dpZtTl66BIA.png",
          "originalHeight": 1624,
          "originalWidth": 3360,
          "strategy": "resample",
          "width": 0,
        },
        "readingTime": 3.8899371069182394,
        "recommends": 5,
        "responsesCreatedCount": 0,
        "slug": "new-year-new-shell",
        "subtitle": "Getting started with fish, the friendly interactive shell",
        "tags": Array [
          Object {
            "name": "Programming",
            "slug": "programming",
          },
          Object {
            "name": "Development",
            "slug": "development",
          },
          Object {
            "name": "Unix",
            "slug": "unix",
          },
        ],
        "title": "New year, new shell",
        "totalClapCount": 44,
        "type": "Post",
        "url": "https://medium.com/@huntie/new-year-new-shell-f62cd3f15fe7",
        "versionId": "e59120a0fc7",
      },
      Object {
        "canonicalUrl": "",
        "coverless": true,
        "createdAt": 1523534597902,
        "creatorId": "e54e8fe3cbbe",
        "detectedLanguage": "en",
        "displayAuthor": "",
        "firstPublishedAt": 1527370662207,
        "inResponseToPostId": "",
        "inResponseToRemovedAt": 0,
        "isNsfw": false,
        "isProxyPost": false,
        "isSeries": false,
        "latestPublishedAt": 1537972719955,
        "latestPublishedVersion": "e5b3d47a80dc",
        "license": 0,
        "previewImage": Object {
          "backgroundSize": "",
          "filter": "",
          "height": 0,
          "imageId": "1*e6USL9jPFm4mjj6J06UhLw.png",
          "originalHeight": 1012,
          "originalWidth": 1238,
          "strategy": "resample",
          "width": 0,
        },
        "readingTime": 5.485849056603774,
        "recommends": 12,
        "responsesCreatedCount": 6,
        "slug": "thoughts-on-the-uk-curve-card",
        "subtitle": "One of the more recent FinTech startups to come out the London tech scene is Curve, a consumer money product launched in January of this…",
        "tags": Array [
          Object {
            "name": "Fintech",
            "slug": "fintech",
          },
          Object {
            "name": "Debit Card",
            "slug": "debit-card",
          },
          Object {
            "name": "Credit Cards",
            "slug": "credit-cards",
          },
          Object {
            "name": "Mobile Wallets",
            "slug": "mobile-wallets",
          },
          Object {
            "name": "Money",
            "slug": "money",
          },
        ],
        "title": "Thoughts on the UK Curve Card",
        "totalClapCount": 106,
        "type": "Post",
        "url": "https://medium.com/@huntie/thoughts-on-the-uk-curve-card-b4125a97e3c3",
        "versionId": "e5b3d47a80dc",
      },
      Object {
        "canonicalUrl": "",
        "coverless": true,
        "createdAt": 1518284342951,
        "creatorId": "e54e8fe3cbbe",
        "detectedLanguage": "en",
        "displayAuthor": "",
        "firstPublishedAt": 1518445389046,
        "inResponseToPostId": "",
        "inResponseToRemovedAt": 0,
        "isNsfw": false,
        "isProxyPost": false,
        "isSeries": false,
        "latestPublishedAt": 1519674751946,
        "latestPublishedVersion": "19a69b6fc9b2",
        "license": 0,
        "previewImage": Object {
          "backgroundSize": "",
          "filter": "",
          "height": 0,
          "imageId": "1*SdpIvSyGVX3LlB5TjVylPQ.png",
          "originalHeight": 1080,
          "originalWidth": 1920,
          "strategy": "resample",
          "width": 0,
        },
        "readingTime": 2.2320754716981135,
        "recommends": 15,
        "responsesCreatedCount": 2,
        "slug": "ia-writer-for-windows-first-impressions",
        "subtitle": "The original focused writing app comes to Windows",
        "tags": Array [
          Object {
            "name": "Writing",
            "slug": "writing",
          },
          Object {
            "name": "Windows",
            "slug": "windows",
          },
          Object {
            "name": "Ia Writer",
            "slug": "ia-writer",
          },
          Object {
            "name": "Apps",
            "slug": "apps",
          },
          Object {
            "name": "Kickstarter",
            "slug": "kickstarter",
          },
        ],
        "title": "iA Writer for Windows first impressions",
        "totalClapCount": 34,
        "type": "Post",
        "url": "https://medium.com/@huntie/ia-writer-for-windows-first-impressions-289a7cdd58e6",
        "versionId": "19a69b6fc9b2",
      },
      Object {
        "canonicalUrl": "",
        "coverless": true,
        "createdAt": 1515325074962,
        "creatorId": "e54e8fe3cbbe",
        "detectedLanguage": "en",
        "displayAuthor": "",
        "firstPublishedAt": 1515698055343,
        "inResponseToPostId": "",
        "inResponseToRemovedAt": 0,
        "isNsfw": false,
        "isProxyPost": false,
        "isSeries": false,
        "latestPublishedAt": 1553352062436,
        "latestPublishedVersion": "a9a6fabdd4d0",
        "license": 0,
        "previewImage": Object {
          "backgroundSize": "",
          "filter": "",
          "height": 0,
          "imageId": "1*AifpzFxsXggIzG-Q_oqO8Q.png",
          "originalHeight": 1070,
          "originalWidth": 1932,
          "strategy": "resample",
          "width": 0,
        },
        "readingTime": 5.676729559748427,
        "recommends": 407,
        "responsesCreatedCount": 10,
        "slug": "10-essential-vim-plugins-for-2018",
        "subtitle": "Vim remains a powerful and ubiquitous application to tackle any number of text editing tasks conveniently from the terminal (some learning…",
        "tags": Array [
          Object {
            "name": "Vim",
            "slug": "vim",
          },
          Object {
            "name": "Vim Plugins",
            "slug": "vim-plugins",
          },
          Object {
            "name": "Top 10",
            "slug": "top-10",
          },
          Object {
            "name": "Unix",
            "slug": "unix",
          },
          Object {
            "name": "Text Editor",
            "slug": "text-editor",
          },
        ],
        "title": "10 essential Vim plugins for 2018",
        "totalClapCount": 2533,
        "type": "Post",
        "url": "https://medium.com/@huntie/10-essential-vim-plugins-for-2018-39957190b7a9",
        "versionId": "a9a6fabdd4d0",
      },
      Object {
        "canonicalUrl": "",
        "coverless": true,
        "createdAt": 1514029699514,
        "creatorId": "e54e8fe3cbbe",
        "detectedLanguage": "en",
        "displayAuthor": "",
        "firstPublishedAt": 1514417839927,
        "inResponseToPostId": "",
        "inResponseToRemovedAt": 0,
        "isNsfw": false,
        "isProxyPost": false,
        "isSeries": false,
        "latestPublishedAt": 1515874706695,
        "latestPublishedVersion": "f36a797cce2f",
        "license": 0,
        "previewImage": Object {
          "backgroundSize": "",
          "filter": "",
          "height": 0,
          "imageId": "1*vQNSInwyoEr8i8sDXJY4CA.png",
          "originalHeight": 800,
          "originalWidth": 1600,
          "strategy": "resample",
          "width": 0,
        },
        "readingTime": 3.4830188679245286,
        "recommends": 2,
        "responsesCreatedCount": 0,
        "slug": "snapshot-testing-in-python-for-sublime-text-packages",
        "subtitle": "A few months ago I authored a small Sublime Text plugin called Sublime tmux, providing a set of commands to open windows within tmux…",
        "tags": Array [
          Object {
            "name": "Sublime Text",
            "slug": "sublimetext",
          },
          Object {
            "name": "Python",
            "slug": "python",
          },
          Object {
            "name": "Unit Testing",
            "slug": "unit-testing",
          },
          Object {
            "name": "Software Development",
            "slug": "software-development",
          },
        ],
        "title": "Snapshot testing in Python for Sublime Text packages",
        "totalClapCount": 25,
        "type": "Post",
        "url": "https://medium.com/@huntie/snapshot-testing-in-python-for-sublime-text-packages-ef8aaecde945",
        "versionId": "f36a797cce2f",
      },
      Object {
        "canonicalUrl": "",
        "coverless": true,
        "createdAt": 1486239404234,
        "creatorId": "e54e8fe3cbbe",
        "detectedLanguage": "en",
        "displayAuthor": "",
        "firstPublishedAt": 1486631944565,
        "inResponseToPostId": "",
        "inResponseToRemovedAt": 0,
        "isNsfw": false,
        "isProxyPost": false,
        "isSeries": false,
        "latestPublishedAt": 1535649095060,
        "latestPublishedVersion": "844573c19057",
        "license": 0,
        "previewImage": Object {
          "backgroundSize": "",
          "filter": "",
          "height": 0,
          "imageId": "1*ePaMoSu83h_HoWz4Pj1o8w.png",
          "originalHeight": 1446,
          "originalWidth": 2944,
          "strategy": "resample",
          "width": 0,
        },
        "readingTime": 3.609433962264151,
        "recommends": 14,
        "responsesCreatedCount": 0,
        "slug": "the-ultra-minimalist-static-site-stack",
        "subtitle": "This year I took a fresh look what was needed in my personal website, having decided to publish my articles here on Medium. Along with a…",
        "tags": Array [
          Object {
            "name": "Metalsmith",
            "slug": "metalsmith",
          },
          Object {
            "name": "Netlify",
            "slug": "netlify",
          },
          Object {
            "name": "Static Site",
            "slug": "static-site",
          },
          Object {
            "name": "Npm Scripts",
            "slug": "npm-scripts",
          },
        ],
        "title": "The ultra-minimalist static site stack",
        "totalClapCount": 30,
        "type": "Post",
        "url": "https://medium.com/@huntie/the-ultra-minimalist-static-site-stack-31458ac4f48d",
        "versionId": "844573c19057",
      },
      Object {
        "canonicalUrl": "",
        "coverless": true,
        "createdAt": 1483703513624,
        "creatorId": "e54e8fe3cbbe",
        "detectedLanguage": "en",
        "displayAuthor": "",
        "firstPublishedAt": 1483733066220,
        "inResponseToPostId": "",
        "inResponseToRemovedAt": 0,
        "isNsfw": false,
        "isProxyPost": false,
        "isSeries": false,
        "latestPublishedAt": 1514394793685,
        "latestPublishedVersion": "677830f49db1",
        "license": 0,
        "previewImage": Object {
          "backgroundSize": "",
          "filter": "",
          "height": 0,
          "imageId": "",
          "originalHeight": 0,
          "originalWidth": 0,
          "strategy": "resample",
          "width": 0,
        },
        "readingTime": 2.2037735849056603,
        "recommends": 39,
        "responsesCreatedCount": 0,
        "slug": "representing-a-uuid-as-a-base-62-hash-id-for-short-pretty-urls",
        "subtitle": "UUIDs are awesome — they provide a non-sequential, comprehensively unique identifier that can be generated separately from the source…",
        "tags": Array [
          Object {
            "name": "Programming",
            "slug": "programming",
          },
          Object {
            "name": "PHP",
            "slug": "php",
          },
          Object {
            "name": "Laravel",
            "slug": "laravel",
          },
          Object {
            "name": "Hashids",
            "slug": "hashids",
          },
          Object {
            "name": "Uuids",
            "slug": "uuid",
          },
        ],
        "title": "Representing a UUID as a Base62 Hash ID for short, pretty URLs",
        "totalClapCount": 120,
        "type": "Post",
        "url": "https://medium.com/@huntie/representing-a-uuid-as-a-base-62-hash-id-for-short-pretty-urls-c30e66bf35f9",
        "versionId": "677830f49db1",
      },
    ],
    "user": Object {
      "allowNotes": 1,
      "backgroundImageId": "",
      "bio": "London-based software developer. Building a platform for mental wellbeing at @unmindhealth. https://alexhunt.io/",
      "createdAt": 1462126354657,
      "imageId": "1*iumTueER5jCHl9TgzYra1g.jpeg",
      "isNsfw": false,
      "isQuarantined": false,
      "isWriterProgramEnrolled": true,
      "mediumMemberAt": 0,
      "name": "Alex Hunt",
      "twitterScreenName": "alxhnt",
      "type": "User",
      "userId": "e54e8fe3cbbe",
      "username": "huntie",
    },
  },
}
`;
